### PR TITLE
ADR 022: GQL compliant notifications

### PR DESCRIPTION
**Pull request by robsdedude**
# GQL Summary Statuses
This PR introduces GQL compliant status objects to the `ResultSummary`.

:warning: This feature is still in **PREVIEW**.
It might be changed without following the deprecation policy.

## Deprecations
`SummaryNotificationPosition` has been deprecated in favor of `SummaryInputPosition`.

## New GQL Status Objects
A new property `ResultSummary.gql_status_objects` has been introduced. It returns a sequence of `GqlStatusObject`s. These objects contain information about the execution of the query including notifications as they exist now (see `SummaryNotification`) but also an outcome status like `00000` for "success", `02000` for "no data", and `00001` for "omitted result".

Eventually, this API is planned to supersede the current notifications API.

The GqlStatusObjects will be presented in the following order:

 * A "no data" (`02xxx`) has precedence over a warning;
 * A "warning" (`01xxx`) has precedence over a success.
 * A "success" (`00xxx`) has precedence over anything informational (`03xxx`).

## Notification Filtering
Some status objects are notifications (vendor-specific definition). Those notifications can be identified using `GqlStatusObject.is_notification`. They contain additional information like a `classification`, `severity`, and `position`. Further, those notifications can be configured to be suppressed in the DBMS saving bandwidth not having to send them to the client as well as compute not having to check for those conditions.  
For this, the following driver config options can be used:

 * `notifications_min_severity` (already existing)
 * `notifications_disabled_classifications` (new, same as the existing `notifications_disabled_categories`).
   If both options are provided, they are merged into a single list of disabled classifications/categories.

All filter config options will affect both the current notifications API as well as the notifications among the GQL status objects.

## Polyfilling
There are now two APIs to access notifications: the old notifications API and the new GQL status API. To provide an easy migration path the driver will, depending on if the server is GQL aware or not (i.e., recent or old neo4j version), polyfill the API not supported by the server on a best-effort basis. In particular, GQL status objects have a GQLSTATUS code that has not counterpart in non-GQL-aware neo4j versions. Therefore, the driver will fill the GQLSTATUS with either `01N42` for warning notifications and `03N42` for informational notifications. Further, such status codes might be used by the server in the transition phase towards full GQL compliance. They indicate a lack of a final status code and will likely change in the future.

A list of all possible GQLSTATUS codes is not yet available as the server-side of this change is still very actively being worked on.

## Example Usage
```python
with neo4j.GraphDatabase.driver(
    URI,
    auth=AUTH,
    notifications_disabled_classifications=[
        # Remove this filter entry to get an additional notification
        # about the usage of a cartesian product.
        neo4j.NotificationClassification.PERFORMANCE
    ],
) as driver:
    _, summary, _ = driver.execute_query(
        "MATCH (n:Foo), (m) RETURN n.nope, m"
    )
    assert isinstance(summary, neo4j.ResultSummary)
    for status in summary.gql_status_objects:
        # The GQLSTATUS code of the status object.
        print("GQLSTATUS:", status.gql_status)
        # A description of the status for human consumption.
        print("description:", status.status_description)

        # Whether the status is a notification and can be filtered.
        print("is notification:", status.is_notification)

        # Notification and thus vendor-specific fields.
        # These fields are only meaningful for notifications.
        if status.is_notification:
            # The position in the query that caused the notification.
            print("position:", status.position)
            # The notification's classification.
            # This is the counterpart to `neo4j.NotificationCategory`,
            # however, the term category is already used for a different
            # context in the context of GQL
            print("classification:", status.classification)
            print("unparsed classification:", status.raw_classification)
            # The notification's severity.
            print("severity:", status.severity)
            print("unparsed severity:", status.raw_severity)

        # Any raw extra information provided by the DBMS:
        print("diagnostic record:", status.diagnostic_record)
        print("=" * 80)
```

Which, when run against an empty 5.19 DBMS, gives the following output
```
GQLSTATUS: 02000
description: note: no data
is notification: False
diagnostic record: {'OPERATION': '', 'OPERATION_CODE': '0', 'CURRENT_SCHEMA': '/'}
================================================================================
GQLSTATUS: 01N42
description: One of the labels in your query is not available in the database, make sure you didn't misspell it or that the label is available when you run this statement in your application (the missing label name is: Foo)
is notification: True
position: line: 1, column: 10, offset: 9
classification: NotificationClassification.UNRECOGNIZED
unparsed classification: UNRECOGNIZED
severity: NotificationSeverity.WARNING
unparsed severity: WARNING
diagnostic record: {'OPERATION': '', 'OPERATION_CODE': '0', 'CURRENT_SCHEMA': '/', '_classification': 'UNRECOGNIZED', '_severity': 'WARNING', '_position': {'column': 10, 'offset': 9, 'line': 1}}
================================================================================
GQLSTATUS: 01N42
description: One of the property names in your query is not available in the database, make sure you didn't misspell it or that the label is available when you run this statement in your application (the missing property name is: nope)
is notification: True
position: line: 1, column: 29, offset: 28
classification: NotificationClassification.UNRECOGNIZED
unparsed classification: UNRECOGNIZED
severity: NotificationSeverity.WARNING
unparsed severity: WARNING
diagnostic record: {'OPERATION': '', 'OPERATION_CODE': '0', 'CURRENT_SCHEMA': '/', '_classification': 'UNRECOGNIZED', '_severity': 'WARNING', '_position': {'column': 29, 'offset': 28, 'line': 1}}
================================================================================
```

***

Depends on:
 * https://github.com/neo4j-drivers/testkit/pull/617

### CODE REVIEW

File: docs/source/api.rst [hunk 1]
```diff
@@ -726,6 +733,27 @@ Notifications are available via :attr:`.ResultSummary.notifications` and :attr:`
...
+
+**This is a preview** (see :ref:`filter-warnings-ref`).
+It might be changed without following the deprecation policy.
+See also
+https://github.com/neo4j/neo4j-python-driver/wiki/preview-features
+
+:Type: :data:`None`, :term:`iterable` of :class:`.NotificationDisabledClassification` and/or :class:`str`
+:Default: :data:`None`
+
+.. versionadded:: 5.21
```
Comment (bigmontz) [inline docs/source/api.rst]:
```suggestion
.. versionadded:: 5.22
```

File: docs/source/api.rst [hunk 2]
```diff
@@ -1197,6 +1233,27 @@ Notifications are available via :attr:`.ResultSummary.notifications` and :attr:`
...
```
Comment (bigmontz) [inline docs/source/api.rst]:
```suggestion
.. versionadded:: 5.22
```

File: docs/source/api.rst [hunk 3]
```diff
@@ -1542,11 +1613,26 @@ NotificationCategory
...
-.. autoclass:: neo4j.SummaryNotificationPosition()
-    :members:
+.. data:: neo4j.SummaryNotificationPosition
+    :annotation: = neo4j.SummaryInputPosition
+
+    Deprecated alias for :class:`.SummaryInputPosition`.
+
+    .. versionadded:: 5.7
+
+    .. versionchanged:: 5.21
```
Comment (bigmontz) [inline docs/source/api.rst]:
```suggestion
    .. versionchanged:: 5.22
```

File: src/neo4j/_api.py [hunk 4]
```diff
@@ -158,9 +169,42 @@ class NotificationDisabledCategory(str, Enum):
...
+    See also
+    https://github.com/neo4j/neo4j-python-driver/wiki/preview-features
+
+    .. seealso::
+        driver config
+        :ref:`driver-notifications-disabled-classifications-ref`,
+        session config
+        :ref:`session-notifications-disabled-classifications-ref`
+
+    .. versionadded:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_api.py]:
```suggestion
    .. versionadded:: 5.22
```

File: src/neo4j/_api.py [hunk 5]
```diff
@@ -210,6 +255,37 @@ class NotificationCategory(str, Enum):
...
+    :attr:`.GqlStatusObject.classification`.
+
+    **This is a preview**.
+    It might be changed without following the deprecation policy.
+    See also
+    https://github.com/neo4j/neo4j-python-driver/wiki/preview-features
+
+    .. seealso:: :attr:`.GqlStatusObject.classification`
+
+    .. versionadded:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_api.py]:
```suggestion
    .. versionadded:: 5.22
```

File: src/neo4j/_meta.py [hunk 6]
```diff
@@ -159,6 +159,10 @@ def preview_warn(message, stack_level=1):
         "See also "
         "https://github.com/neo4j/neo4j-python-driver/wiki/preview-features."
     )
+    # import traceback, sys
+    # print("\n" + "=" * 80)
+    # traceback.print_stack(file=sys.stdout)
+    # print("=" * 80 + "\n")
```
Comment (bigmontz) [inline src/neo4j/_meta.py]:
```suggestion
```

Leftover?

File: src/neo4j/_work/gql_status.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+class GqlStatusObject:
+    """
+    Container for GQLSTATUS information.
+
+    **This is a preview**.
+    It might be changed without following the deprecation policy.
+    See also
+    https://github.com/neo4j/neo4j-python-driver/wiki/preview-features
+
+    .. versionadded:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_work/gql_status.py]:
```suggestion
    .. versionadded:: 5.22
```

File: src/neo4j/_work/gql_status.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    Container additional GQL status information useful for diagnostics.
+
+    The contents are mostly vendor-specific GQL status information.
+
+    **This is a preview**.
+    It might be changed without following the deprecation policy.
+    See also
+    https://github.com/neo4j/neo4j-python-driver/wiki/preview-features
+
+    .. versionadded:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_work/gql_status.py]:
```suggestion
    .. versionadded:: 5.22
```

File: src/neo4j/_work/summary.py [hunk 9]
```diff
@@ -109,21 +128,200 @@ def __init__(self, address: Address, **metadata: t.Any) -> None:
...
+        All following status are notifications like warnings about problematic
+        queries or other valuable information that can be presented in a
+        client.
+
+        **This is a preview** (see :ref:`filter-warnings-ref`).
+        It might be changed without following the deprecation policy.
+        See also
+        https://github.com/neo4j/neo4j-python-driver/wiki/preview-features
+
+        .. versionadded:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py]:
```suggestion
        .. versionadded:: 5.22
```

File: src/neo4j/_work/summary.py [hunk 10]
```diff
@@ -109,21 +128,200 @@ def __init__(self, address: Address, **metadata: t.Any) -> None:
...
+        """
+        # if getattr(self, "_gql_status_objects", None) is not None:
+        #     return self._gql_status_objects
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py]:
```suggestion
```

Leftover?

File: src/neo4j/_work/summary.py [hunk 11]
```diff
@@ -109,21 +128,200 @@ def __init__(self, address: Address, **metadata: t.Any) -> None:
...
+        notification_status_objects = sorted(
+            notification_status_objects,
+            key=status_precedence,
+            reverse=True,
+        )
+        self._gql_status_objects = tuple(notification_status_objects)
+        print([
+            status.gql_status
+            for status in self._gql_status_objects
+        ])
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py]:
```suggestion
```

Debug leftover?

File: src/neo4j/_work/summary.py [hunk 12]
```diff
@@ -216,12 +414,59 @@ def contains_system_updates(self) -> bool:
...
+@dataclass
+class SummaryInputPosition:
+    """Structured form of a gql status/notification position.
+
+    .. seealso::
+        :attr:`.GqlStatusObject.position`,
+        :attr:`.SummaryNotification.position`,
+        :data:`.SummaryNotificationPosition`
+
+    .. versionadded:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py]:
```suggestion
    .. versionadded:: 5.22
```

File: src/neo4j/_work/summary.py [hunk 13]
```diff
@@ -216,12 +414,59 @@ def contains_system_updates(self) -> bool:
...
+    def __str__(self) -> str:
+        return (f"line: {self.line}, column: {self.column}, "
+                f"offset: {self.offset}")
+
+
+# Deprecated alias for :class:`.SummaryInputPosition`.
+#
+# .. versionadded:: 5.7
+#
+# .. versionchanged:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py]:
```suggestion
# .. versionchanged:: 5.22
```

File: src/neo4j/_work/summary.py [hunk 14]
```diff
@@ -276,48 +541,360 @@ def _from_metadata(cls, metadata):
...
+    driver config
+    :ref:`driver-notifications-disabled-classifications-ref` and
+    :ref:`driver-notifications-min-severity-ref` as well as
+    session config
+    :ref:`session-notifications-disabled-classifications-ref` and
+    :ref:`session-notifications-min-severity-ref`.
+
+    .. seealso:: :attr:`.ResultSummary.gql_status_objects`
+
+    .. versionadded:: 5.21
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py]:
```suggestion
    .. versionadded:: 5.22
```

File: src/neo4j/_work/summary.py [hunk 15]
```diff
@@ -276,48 +541,360 @@ def _from_metadata(cls, metadata):
...
+        if hasattr(self, "_severity"):
+            return self._severity
+
+        self._severity = _SEVERITY_LOOKUP.get(
+            self.raw_severity, NotificationSeverity.UNKNOWN
+        )
+        return self._severity
+
+    @property
+    def diagnostic_record(self) -> t.Dict[str, t.Any]:
```
Comment (bigmontz) [inline src/neo4j/_work/summary.py:889]: Do we a have easy way of getting a string repr of this?

File: testkitbackend/_async/requests.py [hunk 16]
```diff
@@ -118,14 +118,14 @@ async def NewDriver(backend, data):
             backend.client_cert_providers[client_cert_provider_id]
         data.mark_item_as_read_if_equals("clientCertificate", None)
         expected_warnings.append(
-            (neo4j.PreviewWarning, "Mutual TLS is a preview feature.")
+            (neo4j.PreviewWarning, r"Mutual TLS is a preview feature\.")
```
Comment (bigmontz) [inline testkitbackend/_async/requests.py:121]: what does r means?

**Review by bigmontz — CHANGES_REQUESTED** [review]
the version is wrong and there are some leftovers

File: testkitbackend/_async/requests.py (hunk 16 above)
Comment (robsdedude) [reply testkitbackend/_async/requests.py:121]: `r""` strings are raw strings. This means that the `\` should not be treated as an escape character but simply as a backslash.

File: src/neo4j/_work/summary.py (hunk 15 above)
Comment (robsdedude) [reply src/neo4j/_work/summary.py:889]: Yes we do. Just throw it into the `str` method (or `print` which does this implicitly).

```python
example_diag_record = {
    "null": None,
    "bool": True,
    "int": 42,
    "float": 3.14,
    "str": "Hello, World!",
    "list": [1, "abc", None],
    "dict": {"key": "value"},
    "point": neo4j.spatial.CartesianPoint((1.1, 2.3, 0.9)),
    "datetime": neo4j.time.DateTime.now(),
    "node": some_node,
    "relationship": some_rel,
}
print("Example diagnostic record:", example_diag_record)
```
will leave you with a lovely
```
Example diagnostic record: {'null': None, 'bool': True, 'int': 42, 'float': 3.14, 'str': 'Hello, World!', 'list': [1, 'abc', None], 'dict': {'key': 'value'}, 'point': POINT(1.1 2.3 0.9), 'datetime': neo4j.time.DateTime(2024, 6, 14, 11, 32, 25, 226660651), 'node': <Node element_id='4:2bcc24d0-d5ef-4540-b0f1-17e0f5d99fb0:0' labels=frozenset({'Foo'}) properties={'hello': 'world'}>, 'relationship': <Relationship element_id='5:2bcc24d0-d5ef-4540-b0f1-17e0f5d99fb0:0' nodes=(<Node element_id='4:2bcc24d0-d5ef-4540-b0f1-17e0f5d99fb0:0' labels=frozenset({'Foo'}) properties={'hello': 'world'}>, <Node element_id='4:2bcc24d0-d5ef-4540-b0f1-17e0f5d99fb0:1' labels=frozenset() properties={}>) type='BAZ' properties={'sice': 1999}>}
```

**Review by bigmontz — APPROVED** [review]
🛎️

**Comment by akollegger** [comment]
Does GQL require an api with "GQL" in all the names?

**Pull request merged**